File: app/src/main/java/ru/santaev/clipboardtranslator/ui/adapter/HistoryAdapter.java
package ru.santaev.clipboardtranslator.ui.adapter;


import android.databinding.DataBindingUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

import ru.santaev.clipboardtranslator.R;
import ru.santaev.clipboardtranslator.databinding.HistoryItemLayoutBinding;
import ru.santaev.clipboardtranslator.db.entity.Translation;

public class HistoryAdapter extends RecyclerView.Adapter<HistoryAdapter.HistoryViewHolder>{

    public interface OnHistoryItemClickedListener{
        void onClick(Translation translation);
    }

    private List<TranslateView> translations;
    private OnHistoryItemClickedListener listener;

    public HistoryAdapter(List<Translation> translations) {
        setHasStableIds(true);
        setTranslations(translations);
    }

    public void setTranslations(List<Translation> translations) {
        this.translations = new ArrayList<>();
        if (translations == null){
            return;
        }
        for (Translation translation : translations) {
            TranslateView translateView = new TranslateView(translation);
            this.translations.add(translateView);
        }
    }

    public Translation getTranslation(int pos) {
        return translations.get(pos).translation;
    }

    public OnHistoryItemClickedListener getListener() {
        return listener;
    }

    public void setListener(OnHistoryItemClickedListener listener) {
        this.listener = listener;
    }

    @Override
    public long getItemId(int position) {
        return null == translations ? 0 : translations.get(position).translation.getId();
    }

    @Override
    public HistoryViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        HistoryItemLayoutBinding binding = DataBindingUtil.inflate(LayoutInflater.from(parent.getContext()),
                R.layout.history_item_layout, parent, false);
        HistoryViewHolder viewHolder = new HistoryViewHolder(binding);
        viewHolder.binding.setIsExpanded(false);
        return viewHolder;
    }

    @Override
    public void onBindViewHolder(HistoryViewHolder holder, int position) {
        TranslateView model = translations.get(position);
        holder.binding.originText.setText(model.translation.getTextSource());
        holder.binding.targetText.setText(model.translation.getTextTarget());
        holder.binding.lang.setText(buildLangText(model.translation));

        // TODO: 20.09.2017 broke childs size in recycler view
        /*if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN){
            ((ViewGroup) holder.binding.getRoot()).getLayoutTransition()
                    .disableTransitionType(LayoutTransition.CHANGING);
        }*/

        holder.binding.setIsExpanded(model.isExpanded);
        holder.binding.executePendingBindings();

        holder.binding.getRoot().setOnClickListener(v -> {
            model.isExpanded = !model.isExpanded;
            holder.binding.setIsExpanded(model.isExpanded);

            if (listener != null){
                listener.onClick(model.translation);
            }
        });

    }

    @Override
    public int getItemCount() {
        return translations == null ? 0 : translations.size();
    }

    private String buildLangText(Translation translation){
        return translation.getLangSource().toUpperCase() +
                "-" + translation.getLangTarget().toUpperCase();
    }

    class TranslateView {

        public Translation translation;
        public boolean isExpanded;

        public TranslateView(Translation translation) {
            this.translation = translation;
            this.isExpanded = false;
        }
    }

    class HistoryViewHolder extends RecyclerView.ViewHolder{

        public HistoryItemLayoutBinding binding;

        public HistoryViewHolder(HistoryItemLayoutBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }
}

File: app/src/main/java/ru/santaev/clipboardtranslator/ui/adapter/HistoryAdapter.kt
package ru.santaev.clipboardtranslator.ui.adapter


import android.databinding.DataBindingUtil
import android.graphics.Canvas
import android.support.v7.widget.CardView
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.helper.ItemTouchHelper
import android.view.LayoutInflater
import android.view.ViewGroup
import com.example.santaev.domain.dto.TranslationDto
import ru.santaev.clipboardtranslator.R
import ru.santaev.clipboardtranslator.databinding.HistoryItemLayoutBinding
import java.util.*

class HistoryAdapter(translations: List<TranslationDto>?) : RecyclerView.Adapter<HistoryAdapter.HistoryViewHolder>() {

    private var translations: MutableList<TranslateView>? = null
    var listener: ((TranslationDto) -> Unit)? = null

    init {
        setHasStableIds(true)
        setTranslations(translations)
    }

    fun setTranslations(translations: List<TranslationDto>?) {
        this.translations = ArrayList()
        if (translations == null) {
            return
        }
        translations
                .map { TranslateView(it) }
                .forEach { this.translations?.add(it) }
    }

    fun getTranslation(pos: Int): TranslationDto {
        return translations!![pos].translation
    }

    override fun getItemId(position: Int): Long {
        return if (null == translations) 0 else translations!![position].translation.id
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): HistoryViewHolder {
        val binding = DataBindingUtil.inflate<HistoryItemLayoutBinding>(LayoutInflater.from(parent.context),
                R.layout.history_item_layout, parent, false)
        val viewHolder = HistoryViewHolder(binding)
        viewHolder.binding.isExpanded = false
        return viewHolder
    }

    override fun onBindViewHolder(holder: HistoryViewHolder, position: Int) {
        val model = translations?.get(position) ?: return
        holder.binding.originText.text = model.translation.sourceText
        holder.binding.targetText.text = model.translation.targetText
        holder.binding.lang.text = buildLangText(model.translation)

        holder.binding.isExpanded = model.isExpanded
        holder.binding.executePendingBindings()

        holder.binding.root.setOnClickListener {
            model.isExpanded = !model.isExpanded
            holder.binding.isExpanded = model.isExpanded
            listener?.invoke(model.translation)
        }

    }

    override fun getItemCount(): Int {
        return if (translations == null) 0 else translations!!.size
    }

    private fun buildLangText(translation: TranslationDto): String {
        return translation.run {
            "${sourceLangCode.toUpperCase()}-${targetLangCode.toUpperCase()}"
        }
    }

    inner class HistoryViewHolder(var binding: HistoryItemLayoutBinding) : RecyclerView.ViewHolder(binding.root)

    class RecyclerViewCallback(
            private val onSwipedListener: (Int) -> Unit
    ) : ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.LEFT) {

        override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
            val position = viewHolder.layoutPosition
            onSwipedListener(position)
        }

        override fun onMove(
                recyclerView: RecyclerView,
                viewHolder: RecyclerView.ViewHolder,
                target: RecyclerView.ViewHolder
        ): Boolean {
            return false
        }

        override fun onSelectedChanged(viewHolder: RecyclerView.ViewHolder?, actionState: Int) {
            super.onSelectedChanged(viewHolder, actionState)
            if (viewHolder != null) {
                val foregroundView = viewHolder.foreground
                ItemTouchHelper.Callback.getDefaultUIUtil().onSelected(foregroundView)
            }
        }

        override fun onChildDrawOver(
                c: Canvas,
                recyclerView: RecyclerView,
                viewHolder: RecyclerView.ViewHolder,
                dX: Float,
                dY: Float,
                actionState: Int,
                isCurrentlyActive: Boolean
        ) {
            val item = (viewHolder as HistoryViewHolder).binding
            item.icDelete.apply {
                val width = item.root.width
                rotation = if (width * 0.5 > -dX) {
                    (width * 0.5F + dX) / width * 45
                } else {
                    0F
                }
            }
            val foregroundView = viewHolder.foreground
            ItemTouchHelper.Callback.getDefaultUIUtil().onDrawOver(c, recyclerView, foregroundView, dX, dY, actionState, isCurrentlyActive)
        }

        override fun clearView(recyclerView: RecyclerView?, viewHolder: RecyclerView.ViewHolder) {
            val foregroundView = viewHolder.foreground
            ItemTouchHelper.Callback.getDefaultUIUtil().clearView(foregroundView)
        }

        override fun onChildDraw(
                c: Canvas,
                recyclerView: RecyclerView,
                viewHolder: RecyclerView.ViewHolder,
                dX: Float,
                dY: Float,
                actionState: Int,
                isCurrentlyActive: Boolean
        ) {
            val foregroundView = viewHolder.foreground
            ItemTouchHelper.Callback.getDefaultUIUtil().onDraw(c, recyclerView, foregroundView, dX, dY,
                    actionState, isCurrentlyActive)
        }

        private val RecyclerView.ViewHolder.foreground: CardView
            get() = (this as HistoryViewHolder).binding.foregroundView
    }

    internal inner class TranslateView(var translation: TranslationDto) {
        var isExpanded: Boolean = false

        init {
            this.isExpanded = false
        }
    }
}
